feat: Update Menu styling to match Fiori3 requirements

salarenko opened the pull request:
#### Please provide a link to the associated issue.

#### Please provide a brief summary of this pull request.
This PR:
Closes: #1265
Closes: #2315
Closes: #2246
Closes: #2672

- Introduces new implementation of Menu component
    - Adds integrated popover
    - Adds menu nesting
    - Adds menu mobile mode
    - Adds integrated keyboard support
    - Adds `fdMenuTrigger` directive to attach menu to other elements
    - Adds menu shortcuts
    - Adds menu before/after addons
    - Moves menu separator to separate component
    - Removes menu group
- Refactors Split Button component to match Fiori3
   - removes `SplitButtonLoadActionTitle`
    - removes `SplitButtonMenuDirective`
- Adds custom `[trigger]` input to PopoverDirective
- Adds `(loadeed)` output fired after popover content has been instantiated, to PopoverDirective
- Updates components using old Menu implementation
    - ApiComponent (Docs)
    - BreadcrumbComponent
- Updates components examples using old Menu implementation
    - ActionBar (Docs)
    - Toolbar (Docs helper)
    - TreeChildComponent
- Changes components Docs examples from using Menu to List
    - Dropdown
    - Popover
    - Tile

***Breaking changes***

**Menu**
* Menu has been migrated from simple list markup to component with integrated popover and other logic
* Menu should not be used for creating list template embedded in other components. To create lists templates use `ListComponent`
* `fd-menu-group` has been removed
* `fd-menu-list` has been removed - menu list is now integrated with `fd-menu`
* `fd-menu-item` can be used only on `li` elements
* `fd-menu-item-addon` has been renamed to `fd-menu-addon`
* `fd-menu-addon` should be used on menu element level to add certain icon
* `Menu[separator]="true"` input has been removed. Use `fd-menu-separator` element to create separators.
* `fd-menu-title` should be used to wrap the text value of menu element

**Split Button**
* `fd-split-button-menu` directive has been removed
* Following inputs: `triggers`, `closeOnOutsideClick`, `closeOnEscapeKey`, `focusTrapped`, `fillControlMode`, `isOpen`, `isOpenChange` have been removed. Use API exposed by `MenuComponent` to configure and control popover.
* `(primaryButtonClicked)` emits `MouseEvent` instead of `void`

**ShellbarUserMenuComponent**
* Removed `noArrow` input
* Removed `addContainerClass` input
* Removed `isDropdown` input
* Removed `appendTo` input
* Removed `isOpen` input/output use ShellbarUserMenuComponent.menu.open()/.close()
* Removed `additionalClasses` input

**ProductMenuComponent**
* Removed `noArrow` input
* Removed `addContainerClass` input
* Removed `isDropdown` input
* Removed `appendTo` input
* Removed `isOpen` input/output use ProductMenuComponent.menu.open()/.close()
* Removed `additionalClasses` input

#### Please check whether the PR fulfills the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/master/CONTRIBUTING.md
- [x] tests for the changes that have been done
- [x] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist

Documentation checklist:
- [x] update `README.md`
- [x] [Breaking Changes Wiki](https://github.com/SAP/fundamental-ngx/wiki/Breaking-Changes)
- [x] Documentation Examples
- [x] Stackblitz works for all examples

InnaAtanasova commented on libs/core/src/lib/menu/menu-item/menu-item.component.ts:
```diff
@@ -0,0 +1,231 @@
+    exportAs: 'fd-menu-item',
+    templateUrl: './menu-item.component.html',
+    changeDetection: ChangeDetectionStrategy.OnPush,
+    host: {
```
missing encapsulation

InnaAtanasova commented on libs/core/src/lib/menu/menu-mobile/menu-mobile/menu-mobile.component.ts:
```diff
@@ -0,0 +1,117 @@
+
+@Component({
+    selector: 'fd-menu-mobile',
+    templateUrl: './menu-mobile.component.html'
```
missing encapsulation and changeDetection

InnaAtanasova commented:
On RTL the "back" arrow in the bar does not change

![Screen Shot 2020-06-04 at 10 10 29 PM](https://user-images.githubusercontent.com/39598672/83829294-8855ca00-a6b0-11ea-8a16-8fcb590ca004.png)

InnaAtanasova commented:
Minor stuff but overall really great job with the component!

salarenko commented on libs/core/src/lib/menu/directives/menu-interactive.directive.spec.ts:
```diff
@@ -0,0 +1,66 @@
+        expect(directive).toBeTruthy();
+    });
+
+    it('should create an instance', () => {
+        expect(directive).toBeTruthy();
+    });
+
+    it('should create an instance', () => {
+        expect(directive).toBeTruthy();
+    });
```
Thanks, I actually missed that one

salarenko commented on libs/core/src/lib/menu/directives/menu-interactive.directive.ts:
```diff
@@ -0,0 +1,58 @@
+    readonly fdMenuLinkClass: boolean = true;
+
+    /** @hidden */
+    constructor(public elementRef: ElementRef) { }
```
Yes, `MenuItem` sets event listeners on`MenuIteractive` and uses his `ElementRef` to easily access its HTML

salarenko commented on libs/core/src/lib/menu/menu.component.ts:
```diff
@@ -8,16 +41,279 @@ import { Component, HostBinding, ViewEncapsulation, Input, ChangeDetectionStrate
+    /** @hidden Sets listeners based on triggers array */
+    listenOnTriggerEvents(): void {
+        if (Array.isArray(this.triggers)) {
+            this.triggers.forEach(trigger => {
```
This doesn't need explicite typing. I can add type annotation there but it will make the code more complex for no reason.

salarenko commented on libs/core/src/lib/menu/menu-separator/menu-separator.component.ts:
```diff
@@ -0,0 +1,12 @@
+
+@Component({
+  selector: 'fd-menu-separator',
+  template: ''
```
Because its used as custom element

salarenko commented on libs/core/src/lib/menu/menu-mobile/menu-mobile/menu-mobile.component.ts:
```diff
@@ -0,0 +1,117 @@
+    private _listenOnMenuOpenChange(): void {
+        this._subscriptions.add(
+            this._menuComponent.isOpenChange
+                .subscribe(isOpen => isOpen ? this._openDialog() : this.dialogRef.close())
```
This doesn't need explicite typing. I can add type annotation there but it will make the code more complex for no reason.

salarenko commented on apps/docs/src/app/core/component-docs/action-bar/examples/action-bar-contextual-menu-example.component.html:
```diff
@@ -2,39 +2,44 @@
+                [compact]="true"
+            ></button>
+
+            <fd-menu #menu="fdMenu" placement="bottom-end" [compact]="true">
```
This is not only allowed, this is officially recommended :stuck_out_tongue:
But you are right, I'll revert it for consistency

InnaAtanasova replied:
I read this yesterday :) Maybe create an issue and update the whole documentation

InnaAtanasova commented:
@salarenko missing Storybook

mikerodonnell89 commented:
Mobile menu examples in the spec have a footer cancel button, rather than the top-right close X we have here

mikerodonnell89 commented:
The full screen mobile menu should not have the rounded borders on the corners.  I double checked with the designers on this

mikerodonnell89 commented:
Should there be a way to open/close programmatically?

salarenko commented:
@InnaAtanasova 
- I create story book page for menu

@mikerodonnell89 
- I changed default configuration to the dialog with "Cancel" button
- There is a ticket for removing rounded borders for mobile dialogs in Fundamental Styles

@stefanoScalzo 
- I implemented this hover behavior based on [SAPUI5](https://sapui5.hana.ondemand.com/#/entity/sap.ui.unified.Menu/sample/sap.ui.unified.sample.MenuItemEventing)
- Could you be more specific what is the problem?

JKMarkowski commented on libs/core/src/lib/menu/directives/menu-shortcut.directive.ts:
```diff
@@ -0,0 +1,50 @@
+
+    /** @hidden */
+    private _hideOnMobile(): void {
+        this._menuItem.menuService.isMobileMode
```
Could you add subscription guard there?

JKMarkowski commented on libs/core/src/lib/popover/popover-directive/popover.directive.ts:
```diff
@@ -48,6 +48,10 @@ export class PopoverDirective implements OnInit, OnDestroy, OnChanges {
     @Input()
     isOpen: boolean = false;
 
+    /**  */
+    @Input()
+    popoverTrigger?: ElementRef;
+
```
Let's add some more describtive comment

JKMarkowski commented on libs/core/src/lib/shellbar/product-menu/product-menu.component.ts:
```diff
@@ -1,16 +1,8 @@
 import { MenuComponent } from '../../menu/menu.component';
 import { ShellbarMenuItem } from '../model/shellbar-menu-item';
+import { Placement } from 'popper.js';
+import { PopoverFillMode } from '../../..';
```
Let's put direct import

stefanoScalzo commented:
Focus still moves on hover

salarenko commented:
@stefanoScalzo Thanks for your comprehensive review. Could you catch up with the conversation and respond to the question?
